### PR TITLE
Feature/ensemble

### PR DESCRIPTION
**Pull request by arbennett**
This pull request adds the `Ensemble` class, which allows for running multiple SUMMA simulations with varying decision and/or local parameter values. The ensemble cannot yet run ensembles with differing initial conditions or local attribute values. To enable those further work on writing the modified NetCDF inputs to SUMMA would have to be done.

Sample usage:
```python
import pysumma as ps


param_options = {
    'fixedThermalCond_snow': np.arange(0.35, 0.85, 0.05)
}
config = ps.ensemble.parameter_product(param_options)

# initialize and start runs
e = ps.Ensemble(summa_exe, file_manager, config, num_workers=len(config))
e.start('local', arg_list=['export OMP_NUM_THREADS=1', 
                           'export LD_LIBRARY_PATH=/opt/local/lib'])
e.monitor()

# check status
for i, (n, s) in enumerate(e.simulations.items()):
    if s._status != 'Success':
        print(i, f'{n:100}', s._status)

# open output
ds = e.merge_output()
```

### CODE REVIEW

File: pysumma/simulation.py [hunk 1]
```diff
@@ -41,190 +40,74 @@ def __init__(self, executable, filemanager, run_suffix='_pysumma_run'):
...
-            out_file_path = (
-                self.manager.output_path.value
-                + self.manager.output_prefix.value + '_output_'
-                + run_suffix + '_timestep.nc')
-        return out_file_path
+        run_cmd = self.gen_summa_cmd(processes, prerun_cmds,
+                                     startGRU, countGRU, iHRU,
+                                     freq_restart, progress)
+
+        fman_dir = os.path.dirname(self.manager_path.value)
```
Comment (DavidChoi76) [inline pysumma/simulation.py]: self.manager_path.value -> self.manager_path
There is no value in manager_path.

File: pysumma/simulation.py [hunk 2]
```diff
@@ -41,190 +40,74 @@ def __init__(self, executable, filemanager, run_suffix='_pysumma_run'):
...
-            if 'FATAL ERROR' in output:
-                raise Exception("SUMMA failed to execute!")
-            # define output file name as sopron version of summa
-            out_file_path = (self.manager.output_path.value
-                             + self.manager.output_prefix.value + '_output_'
-                             + run_suffix + '_timestep.nc')
+    def gen_summa_cmd(self, processes=1, prerun_cmds=[],
+                  startGRU=None, countGRU=None, iHRU=None, freq_restart=None,
+                  progress='m'):
+        prerun_cmds.append('export OMP_NUM_THREADS={}'.format(processes))
```
Comment (DavidChoi76) [inline pysumma/simulation.py:46]: Even though I didn't use `prerun_cmds`, it showed up in cmd command for SUMMA. So if there is `export OMP_NUM_THREADS=' inside SUMMA cmd, an error occurred.
Change to 
if prerun_cmds != []:  
            prerun_cmds.append('export OMP_NUM_THREADS={}'.format(processes))

File: pysumma/simulation.py [hunk 3]
```diff
@@ -25,12 +25,11 @@ class Simulation(object):
     local_attributes: xr.Dataset = None
     parameter_trial: xr.Dataset = None
 
-    def __init__(self, executable, filemanager, run_suffix='_pysumma_run'):
+    def __init__(self, executable, filemanager):
```
Comment (DavidChoi76) [inline pysumma/simulation.py:28]: In here, do you think what variable do we have to define for `executable'? I think if we want to define executable itself such as executable="uwhydro/summa:sopron_2018", we need to delete every "docker_img" variable. If we don't delete it, "docker_img" defines inside some funtions. So there is no meaning to define executable in Simulation class.

File: pysumma/option.py [hunk 4]
```diff
@@ -12,7 +12,7 @@ def __init__(self, name, value=None):
         self.name = name
```
Comment (DavidChoi76) [inline pysumma/option.py:12]: I have a question. What does `optiontype` mean in OptionContainer class?  31 line of option.py

File: pysumma/ensemble.py [hunk 5]
```diff
@@ -0,0 +1,164 @@
+from copy import deepcopy
+from distributed import Client
+import pandas as pd
+import time
+import xarray as xr
+
+from .simulation import Simulation
+from .utils import ChainDict, product_dict
+
+
```
Comment (DavidChoi76) [inline pysumma/ensemble.py:10]: for i, (n, s) in enumerate(e.simulations.items()):
    if s._status != 'Success':
        print(i, f'{n:100}', s._status)

When I ran this code, there was an error. 
        
0 ++fixedThermalCond_snow=0.35++                                                                       Error
1 ++fixedThermalCond_snow=0.39999999999999997++                                                        Error
2 ++fixedThermalCond_snow=0.44999999999999996++                                                        Error
3 ++fixedThermalCond_snow=0.49999999999999994++                                                        Error
4 ++fixedThermalCond_snow=0.5499999999999999++                                                         Error
5 ++fixedThermalCond_snow=0.5999999999999999++                                                         Error
6 ++fixedThermalCond_snow=0.6499999999999999++                                                         Error
7 ++fixedThermalCond_snow=0.7++                                                                        Error
8 ++fixedThermalCond_snow=0.7499999999999999++                                                         Error
9 ++fixedThermalCond_snow=0.7999999999999998++                                                         Error

Comment (DavidChoi76) [reply pysumma/ensemble.py:10]: I reviewed filemanagers using `e.simulations.items()`, but it looked there was no error.

File: pysumma/simulation.py (hunk 1 above)
Comment (arbennett) [reply pysumma/simulation.py]: Thanks, will fix this.

File: pysumma/simulation.py (hunk 2 above)
Comment (arbennett) [reply pysumma/simulation.py:46]: This makes it explicit if SUMMA is being run with multithreading, and I think is correct. I am updating the docker run to fix this.

File: pysumma/simulation.py (hunk 3 above)
Comment (arbennett) [reply pysumma/simulation.py:28]: I well remove the `docker_img` variable throughout. I think it's better to just use `executable`.

File: pysumma/ensemble.py (hunk 5 above)
Comment (arbennett) [reply pysumma/ensemble.py:10]: I'm not sure what is happening without more information.

File: pysumma/option.py (hunk 4 above)
Comment (arbennett) [reply pysumma/option.py:12]: `optiontype` is the type of option. All options are subclasses of the `BaseOption` which is defined in the same file. For example, in `decisions.py` we define `DecisionOption` which specifies what a single option in the decisions file should look like.

File: pysumma/simulation.py (hunk 2 above)
Comment (DavidChoi76) [reply pysumma/simulation.py:46]: <Command>
docker run -v /media/sf_pysumma/a0105d479c334764ba84633c5b9c1c01/a0105d479c334764ba84633c5b9c1c01/data/contents/summaTestCases_2.x/settings/wrrPaperTestCases/figure07:/media/sf_pysumma/a0105d479c334764ba84633c5b9c1c01/a0105d479c334764ba84633c5b9c1c01/data/contents/summaTestCases_2.x/settings/wrrPaperTestCases/figure07 -v /media/sf_pysumma/a0105d479c334764ba84633c5b9c1c01/a0105d479c334764ba84633c5b9c1c01/data/contents/summaTestCases_2.x/settings/:/media/sf_pysumma/a0105d479c334764ba84633c5b9c1c01/a0105d479c334764ba84633c5b9c1c01/data/contents/summaTestCases_2.x/settings/ -v /media/sf_pysumma/a0105d479c334764ba84633c5b9c1c01/a0105d479c334764ba84633c5b9c1c01/data/contents/summaTestCases_2.x/testCases_data/inputData/fieldData/reynolds/:/media/sf_pysumma/a0105d479c334764ba84633c5b9c1c01/a0105d479c334764ba84633c5b9c1c01/data/contents/summaTestCases_2.x/testCases_data/inputData/fieldData/reynolds/ -v /media/sf_pysumma/a0105d479c334764ba84633c5b9c1c01/a0105d479c334764ba84633c5b9c1c01/data/contents/summaTestCases_2.x/output/wrrPaperTestCases/figure07/:/media/sf_pysumma/a0105d479c334764ba84633c5b9c1c01/a0105d479c334764ba84633c5b9c1c01/data/contents/summaTestCases_2.x/output/wrrPaperTestCases/figure07/export OMP_NUM_THREADS=1 && export OMP_NUM_THREADS=1 &&  uwhydro/summa:sopron_2018 -p never -s hs -m /media/sf_pysumma/a0105d479c334764ba84633c5b9c1c01/a0105d479c334764ba84633c5b9c1c01/data/contents/summaTestCases_2.x/settings/wrrPaperTestCases/figure07/summa_fileManager_riparianAspenSimpleResistance.txt
<Error messag>
docker: invalid reference format: repository name must be lowercase.
See 'docker run --help'.

There's still an error here. As I told before, we need if statement for this.
if prerun_cmds != []:
      prerun_cmds.append('export OMP_NUM_THREADS={}'.format(processes))

File: pysumma/ensemble.py (hunk 5 above)
Comment (DavidChoi76) [reply pysumma/ensemble.py:10]: image
image
image

---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
~/anaconda3/envs/pysumma/lib/python3.6/site-packages/distributed/client.py in _gather(self, futures, errors, direct, local_worker)
   1486                         try:
-> 1487                             st = self.futures[key]
   1488                             exception = st.exception

KeyError: '_submit-490b5bd250fdfdaf9c3c2f67410c25a9'

During handling of the above exception, another exception occurred:

CancelledError                            Traceback (most recent call last)
<ipython-input-28-02bf69958f4e> in <module>
      1 e.start('local', arg_list=['export OMP_NUM_THREADS=1', 
      2                            'export LD_LIBRARY_PATH=/opt/local/lib'])
----> 3 e.monitor()

~/anaconda3/envs/pysumma/lib/python3.6/site-packages/pysumma/ensemble.py in monitor(self)
    126 
    127     def monitor(self):
--> 128         simulations = self._client.gather(self.submissions)
    129         for s in simulations:
    130             self.simulations[s.run_suffix] = s

~/anaconda3/envs/pysumma/lib/python3.6/site-packages/distributed/client.py in gather(self, futures, errors, maxsize, direct, asynchronous)
   1654             return self.sync(self._gather, futures, errors=errors,
   1655                              direct=direct, local_worker=local_worker,
-> 1656                              asynchronous=asynchronous)
   1657 
   1658     @gen.coroutine

~/anaconda3/envs/pysumma/lib/python3.6/site-packages/distributed/client.py in sync(self, func, *args, **kwargs)
    674             return future
    675         else:
--> 676             return sync(self.loop, func, *args, **kwargs)
    677 
    678     def __repr__(self):

~/anaconda3/envs/pysumma/lib/python3.6/site-packages/distributed/utils.py in sync(loop, func, *args, **kwargs)
    280             e.wait(10)
    281     if error[0]:
--> 282         six.reraise(*error[0])
    283     else:
    284         return result[0]

~/.local/lib/python3.6/site-packages/six.py in reraise(tp, value, tb)
    691             if value.__traceback__ is not tb:
    692                 raise value.with_traceback(tb)
--> 693             raise value
    694         finally:
    695             value = None

~/anaconda3/envs/pysumma/lib/python3.6/site-packages/distributed/utils.py in f()
    265             if timeout is not None:
    266                 future = gen.with_timeout(timedelta(seconds=timeout), future)
--> 267             result[0] = yield future
    268         except Exception as exc:
    269             error[0] = sys.exc_info()

~/.local/lib/python3.6/site-packages/tornado/gen.py in run(self)
    727 
    728                     try:
--> 729                         value = future.result()
    730                     except Exception:
    731                         exc_info = sys.exc_info()

~/.local/lib/python3.6/site-packages/tornado/gen.py in run(self)
    740                             exc_info = None
    741                     else:
--> 742                         yielded = self.gen.send(value)
    743 
    744                 except (StopIteration, Return) as e:

~/anaconda3/envs/pysumma/lib/python3.6/site-packages/distributed/client.py in _gather(self, futures, errors, direct, local_worker)
   1491                             six.reraise(CancelledError,
   1492                                         CancelledError(key),
-> 1493                                         None)
   1494                         else:
   1495                             six.reraise(type(exception),

~/.local/lib/python3.6/site-packages/six.py in reraise(tp, value, tb)
    691             if value.__traceback__ is not tb:
    692                 raise value.with_traceback(tb)
--> 693             raise value
    694         finally:
    695             value = None

CancelledError: _submit-490b5bd250fdfdaf9c3c2f67410c25a9

File: pysumma/simulation.py [hunk 6]
```diff
@@ -74,30 +74,30 @@ def run_local(self, run_suffix=None, processes=1, prerun_cmds=[],
...
-        cmd = "".join(["docker run -v {}:{}".format(fman_dir, fman_dir),
-                       " -v {}:{}".format(settings_path, settings_path),
-                       " -v {}:{}".format(input_path, input_path),
-                       " -v {}:{} ".format(output_path, output_path),
-                       run_cmd])
+        cmd = ''.join(['docker run -v {}:{}'.format(fman_dir, fman_dir),
+                       ' -v {}:{}'.format(settings_path, settings_path),
+                       ' -v {}:{}'.format(input_path, input_path),
+                       ' -v {}:{}'.format(output_path, output_path),
+                       '/bin/bash -c "',
```
Comment (DavidChoi76) [inline pysumma/simulation.py:92]: '/bin/bash -c "' occurred an error in run command.
We need to delete this.

**Pull request merged**